instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.tracer.ClientSpan;
+import io.opentelemetry.instrumentation.api.tracer.ServerSpan;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+// TODO(anuraaga): Need to define what are actually useful knobs, perhaps even providing a
+// base-class
+// for instrumentation library builders.
+/**
+ * An instrumenter of the start and end of a request/response lifecycle. Almost all instrumentation
+ * of libraries falls into modeling start and end, generating observability signals from these such
+ * as a tracing {@link Span}, or metrics such as the duration taken, active requests, etc. When
+ * instrumenting a library, there will generally be three steps.
+ *
+ * <ul>
+ *   <li>Create an {@link Instrumenter} using {@link InstrumenterBuilder}. Use the builder to
+ *       configure any library-specific customizations, and also expose useful knobs to your user.
+ *   <li>Call {@link Instrumenter#start(Context, Object)} at the beginning of a request.
+ *   <li>Call {@link Instrumenter#end(Context, Object, Object, Throwable)} at the end of a request.
+ * </ul>
+ */
+public class Instrumenter<REQUEST, RESPONSE> {
+
+  /** Returns a new {@link InstrumenterBuilder}. */
+  public static <REQUEST, RESPONSE> InstrumenterBuilder<REQUEST, RESPONSE> newBuilder(
+      OpenTelemetry openTelemetry,
+      String instrumentationName,
+      SpanNameExtractor<? super REQUEST> spanNameExtractor) {
```

Comment:
I would prefer to limit the arguments in this method, maybe just otel and instrumentationName are enough and ask users to set the other properties in the builder?

The main reason is that metrics may also have some special "extractor" etc, so I would say let's just have things that are common (or at least we believe are common at this point between traces and metrics).

Reply:
These are all the required parameters for tracing so they're here - the only default span name we could provide is something like `unknown`, but we'd rather have instrumentation be required to pass it in. Or we could have a "required builder method" but that's not as compile-time-safe.

Reply:
We could make tracing/metrics optional "components" of the instrumenter:

```java
Instrumenter.builder(openTelemetry, instrumentationName)
    .enableTracing(spanNameExtractor)         // whatever's required for tracing
    .enableMetrics(metricsSomethingExtractor) // whatever's required for metrics
    // ...
    .newInstrumeter();
```

This API would be a bit more explicit, but also more elastic and less prone to backwards incompatible changes.

Reply:
@mateuszrzeszutek - Hmm - this is an interesting proposal, would like to think it through more as a followup. One open question is whether we want to make it easy to only enable one signal and not the other.